### ISSUE TITLE
DELETE RETURNING on table with oids is broken

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#8765]
### Greenplum version or build
6X_STABLE. Seems to work on master, but I noticed this while working on the PostgeSQL 9.6 branch, where it was broken again by commit 617db3a2d8.

### Step to reproduce the behavior

```
postgres=# CREATE TABLE t1 (a int, b text) with oids;
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  OIDS=TRUE is not recommended for user-created tables
HINT:  Use OIDS=FALSE to prevent wrap-around of the OID counter.
CREATE TABLE
postgres=# insert into t1 values (1, 'foo');
INSERT 30240 1
postgres=#   DELETE FROM  t1 RETURNING t1.a, t1.b;
 a | b 
---+---
 0 | 
(1 row)

DELETE 1
```
In this example, the returned results are incorrect. With different data or table layout, you can get various errors and segfaults, too.

**Comment by hlinnaka** [comment]
This issue stems from confusion on whether or not the tuples stored in the final tuplestore have OIDs or not. The slot coming from the RETURNING clause has been marked to have OIDs, but it's read back to a slot without oids.

**Comment by hlinnaka** [comment]
@gkokolatos pointed out that this was fixed in the 9.5 merge, by this commit on the merge iteration branch:

```
commit ff8b6c638af31c8343d090a459bafdf802644f51
Author: Pengzhou Tang <ptang@pivotal.io>
Date:   Mon May 6 06:33:04 2019 -0400

    Assign result tupleDescriptor of Motion node from outer plan
    
    Motion node used to assign its result tupleDescriptor from its targetlist
    with ExecAssignResultTypeFromTL(), the tupleDesc of it always been forced
    with option hasoid=true for update/delete statements on tables with oids.
    
    This occurs even it's a update/delete returning statement in which case
    the underlying ModifyTable node alway has a tupleDesc (hasoid=false), so
    a mismatch of tuple descriptor occurs. What's more, for
    PORTAL_ONE_RETURNING strategy, we store its results in the portal's
    tuplestore first with tupleDesc (hasoid = true) and the fetch the
    results from the tuplestore with tupleDesc (hasoid = false), this is
    also incorrect.
    
    Idealy, Motion node should initialize its result tupleDescriptor from
    its outer planstate, however, 9b8f5c0b introduced an optimize in which
    outer planstate of Motion is not initialized in some cases, eg, receiver
    in QEs, so we try outer planstate first and then targetlist. For the
    case shown above, Motion node in QD can get the result tupleDesc from
    its outer planstate.
    
    This fix the crash in regression test rowsecurity.

diff --git a/src/backend/executor/nodeMotion.c b/src/backend/executor/nodeMotion.c
index bdb35ecb049..9cb6d4943aa 100644
--- a/src/backend/executor/nodeMotion.c
+++ b/src/backend/executor/nodeMotion.c
@@ -885,6 +885,7 @@ ExecInitMotion(Motion *node, EState *estate, int eflags)
 	Slice	   *sendSlice = NULL;
 	Slice	   *recvSlice = NULL;
 	SliceTable *sliceTable = estate->es_sliceTable;
+	PlanState  *outerPlan;
 
 #ifdef CDB_MOTION_DEBUG
 	int			i;
@@ -1017,7 +1018,12 @@ ExecInitMotion(Motion *node, EState *estate, int eflags)
 	 * initialize tuple type.  no need to initialize projection info because
 	 * this node doesn't do projections.
 	 */
-	ExecAssignResultTypeFromTL(&motionstate->ps);
+	outerPlan = outerPlanState(motionstate);
+	if (outerPlan && ExecGetResultType(outerPlan))
+		ExecAssignResultType(&motionstate->ps, ExecGetResultType(outerPlan));
+	else
+		ExecAssignResultTypeFromTL(&motionstate->ps);
+
 	motionstate->ps.ps_ProjInfo = NULL;
 	tupDesc = ExecGetResultType(&motionstate->ps);
 
```

**Comment by hlinnaka** [comment]
I was wrong about commit 617db3a, that commit was unrelated. This got broken on the iteration branch, because I reverted the above fix there, because it was causing other troubles.
